src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs

```diff
@@ -148,12 +148,15 @@ protected override SourceText Recover(CancellationToken cancellationToken)
                 }
             }
 
-            protected override Task SaveAsync(SourceText text, CancellationToken cancellationToken)
+            protected override async Task SaveAsync(SourceText text, CancellationToken cancellationToken)
             {
                 Contract.ThrowIfFalse(_storage == null); // Cannot save more than once
 
-                _storage = _parent._storageService.CreateTemporaryTextStorage(CancellationToken.None);
-                return _storage.WriteTextAsync(text);
+                var storage = _parent._storageService.CreateTemporaryTextStorage(cancellationToken);
+                await storage.WriteTextAsync(text).ConfigureAwait(false);
+
+                // make sure write is done before setting _storage field
```

Comment:
The interlocked isn't required, right? Because if the effect of WriteTextAsync isn't observable before it's returned, we've got a lot more bugs than that.